tests/testthat/test-utils-callbacks.R

```diff
@@ -0,0 +1,60 @@
+test_that("CallbackManager catches argument mismatches", {
+  callbacks <- CallbackManager$new()
+
+  expect_snapshot(error = TRUE, {
+    callbacks$add("foo")
+    callbacks$add(function(x, y) x + y)
+    callbacks$add(function(x = 1, y) x + y)
+  })
+
+  expect_silent(callbacks$add(function(x, ...) x))
+  expect_silent(callbacks$add(function(x, y = 1) x + y))
+
+  # Callbacks require one argument, but throw standard R error
+  expect_error(callbacks$invoke())
+  expect_error(callbacks$invoke(1, 2))
+})
+
+test_that("CallbackManager invokes callbacks", {
+  callbacks <- CallbackManager$new()
+  res1 <- NULL
+  res2 <- NULL
+
+  cb1 <- callbacks$add(function(value) {
+    res1 <<- list(value = value, time = Sys.time())
+  })
+
+  cb2 <- callbacks$add(function(...) {
+    res <- list(time = Sys.time())
+    res["value"] <- list(...)
+    res2 <<- res
+  })
+
+  expect_equal(callbacks$count(), 2)
+
+  # Callbacks don't return a value
+  expect_null(callbacks$invoke(list(x = 1, y = 2)))
+
+  # Callbacks receive expected arguments
+  expect_equal(res1$value, list(x = 1, y = 2))
+  expect_equal(res2$value, list(x = 1, y = 2))
+  # Callbacks are invoked in reverse order
+  expect_true(res1$time > res2$time)
+
+  # Unregistering a callback
+  res1_og <- res1
+  cb1()
```

Comment:
I'd recommend testing that this is idempotent, i.e. if you call it again, nothing happens.